StyleCop.Analyzers/StyleCop.Analyzers.Test/Verifiers/DiagnosticVerifier.cs

```diff
@@ -135,71 +135,120 @@ private static void VerifyDiagnosticResults(IEnumerable<Diagnostic> actualResult
         {
             int expectedCount = expectedResults.Count();
             int actualCount = actualResults.Count();
+                StringBuilder errorStringBuilder = new StringBuilder();
 
             if (expectedCount != actualCount)
             {
-                string diagnosticsOutput = actualResults.Any() ? FormatDiagnostics(analyzer, actualResults.ToArray()) : "    NONE.";
-
-                Assert.True(false,
-                    string.Format("Mismatch between number of diagnostics returned, expected \"{0}\" actual \"{1}\"\r\n\r\nDiagnostics:\r\n{2}\r\n", expectedCount, actualCount, diagnosticsOutput));
+                errorStringBuilder.AppendLine($"Mismatch between number of diagnostics returned, expected \"{expectedCount}\" actual \"{actualCount}\"");
             }
 
+            AppendMismatchErrorText(actualResults, analyzer, expectedResults, errorStringBuilder);
+
+            string errorString = errorStringBuilder.ToString();
+            Assert.True(string.IsNullOrEmpty(errorString), errorString);
+
+            //The only thing that could still be wrong is that the duplicated entries in actualResult and expectedResults
+            // don't match. Example:
+            // Let a, b, c be diagnostics and
+            // actual: a, b, b, c
+            // expected: a, b, c, c
+            // Then every actual result was expected and every expected result was reported, but they still don't match.
+            // We shouldn't report multiple diagnostics for a single location so this should not happen, but for the sake
+            // of correctness of this method we should still check it.
             for (int i = 0; i < expectedResults.Length; i++)
             {
                 var actual = actualResults.ElementAt(i);
                 var expected = expectedResults[i];
 
-                if (expected.Line == -1 && expected.Column == -1)
+                AssertDiagnosticsMatch(analyzer, actual, expected);
+            }
+        }
+
+        private static void AppendMismatchErrorText(IEnumerable<Diagnostic> actualResults, DiagnosticAnalyzer analyzer, DiagnosticResult[] expectedResults, StringBuilder errorStringBuilder)
+        {
+            foreach (var actual in actualResults)
+            {
+                bool success = false;
+                foreach (var expected in expectedResults)
                 {
-                    if (actual.Location != Location.None)
+                    try
+                    {
+                        AssertDiagnosticsMatch(analyzer, actual, expected);
```

Comment:
:exclamation: Avoid using exception handling for flow control. In this, it will become impossible to distinguish between `XunitException` which was thrown from this code and `XunitException` which is thrown for a legitimate test failure.